File: app/src/main/java/com/scatl/uestcbbs/module/message/view/PrivateChatActivity.java
package com.scatl.uestcbbs.module.message.view;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.EditText;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.RecyclerView;

import com.luck.picture.lib.PictureSelector;
import com.luck.picture.lib.config.PictureConfig;
import com.luck.picture.lib.config.PictureMimeType;
import com.luck.picture.lib.entity.LocalMedia;
import com.scatl.uestcbbs.R;
import com.scatl.uestcbbs.annotation.ToastType;
import com.scatl.uestcbbs.base.BaseActivity;
import com.scatl.uestcbbs.base.BaseEvent;
import com.scatl.uestcbbs.module.message.MessageManager;
import com.scatl.uestcbbs.widget.MyLinearLayoutManger;
import com.scatl.uestcbbs.widget.emoticon.EmoticonPanelLayout;
import com.scatl.uestcbbs.entity.PrivateChatBean;
import com.scatl.uestcbbs.entity.SendPrivateMsgResultBean;
import com.scatl.uestcbbs.entity.UploadResultBean;
import com.scatl.uestcbbs.helper.glidehelper.GlideEngineForPictureSelector;
import com.scatl.uestcbbs.module.message.adapter.PrivateChatAdapter;
import com.scatl.uestcbbs.module.message.presenter.PrivateChatPresenter;
import com.scatl.uestcbbs.util.Constant;
import com.scatl.uestcbbs.util.ImageUtil;
import com.scatl.uestcbbs.util.ToastUtil;

import org.greenrobot.eventbus.EventBus;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import am.widget.smoothinputlayout.SmoothInputLayout;

public class PrivateChatActivity extends BaseActivity<PrivateChatPresenter> implements PrivateChatView{

    private Toolbar toolbar;
    private RecyclerView recyclerView;
    private PrivateChatAdapter privateChatAdapter;
    private ImageView addImageBtn, addEmoticonBtn, senBtn;
    private EditText chatContent;
    private EmoticonPanelLayout emoticonPanelLayout;
    private SmoothInputLayout lytContent;

    private int hisId;
    private String hisName, sendType, sendContent;
    private boolean isNewPm = false;

    @Override
    protected void getIntent(Intent intent) {
        super.getIntent(intent);
        hisId = intent.getIntExtra(Constant.IntentKey.USER_ID, Integer.MAX_VALUE);
        hisName = intent.getStringExtra(Constant.IntentKey.USER_NAME);
        isNewPm = intent.getBooleanExtra(Constant.IntentKey.IS_NEW_PM, false);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    protected int setLayoutResourceId() {
        return R.layout.activity_private_chat;
    }

    @Override
    protected void findView() {
        toolbar = findViewById(R.id.toolbar);
        recyclerView = findViewById(R.id.private_chat_rv);
        addImageBtn = findViewById(R.id.private_chat_add_photo);
        addEmoticonBtn = findViewById(R.id.private_chat_add_emoticon);
        chatContent = findViewById(R.id.private_chat_edittext);
        senBtn = findViewById(R.id.private_chat_send_btn);
        emoticonPanelLayout = findViewById(R.id.private_chat_emoticon_layout);
        lytContent = findViewById(R.id.sil_lyt_content);
    }

    @Override
    protected void initView() {
        super.initView();
        toolbar.setTitle(hisName);

        senBtn.setOnClickListener(this);
        addImageBtn.setOnClickListener(this);
        addEmoticonBtn.setOnClickListener(this::onClickListener);
        chatContent.setOnClickListener(this::onClickListener);

        privateChatAdapter = new PrivateChatAdapter(R.layout.item_private_chat);
        privateChatAdapter.setHasStableIds(true);
        MyLinearLayoutManger myLinearLayoutManger = new MyLinearLayoutManger(this);
        myLinearLayoutManger.setStackFromEnd(true);
        recyclerView.setLayoutManager(myLinearLayoutManger);
        recyclerView.setAdapter(privateChatAdapter);
        recyclerView.setLayoutAnimation(AnimationUtils.loadLayoutAnimation(this, R.anim.layout_animation_scale_in));

        presenter.getPrivateMsg(hisId, this);
    }

    @Override
    protected PrivateChatPresenter initPresenter() {
        return new PrivateChatPresenter();
    }

    @Override
    protected void onClickListener(View view) {

        if (view.getId() == R.id.private_chat_add_photo) {
            PictureSelector.create(this)
                    .openGallery(PictureMimeType.ofImage())
                    .isCamera(true)
                    .isGif(false)
                    .showCropFrame(false)
                    .hideBottomControls(false)
                    .theme(com.luck.picture.lib.R.style.picture_WeChat_style)
                    .maxSelectNum(1)
                    .isEnableCrop(false)
                    .imageEngine(GlideEngineForPictureSelector.createGlideEngine())
                    .forResult(PictureConfig.CHOOSE_REQUEST);
        }

        if (view.getId() == R.id.private_chat_add_emoticon) {
            if (emoticonPanelLayout.getVisibility() == View.GONE) {
                lytContent.closeKeyboard(true);// 关闭键盘
                lytContent.showInputPane(true);//显示面板
            } else {
                lytContent.closeInputPane();// 关闭面板
                lytContent.showKeyboard();// 显示键盘
            }
        }

        if (view.getId() == R.id.private_chat_edittext) {
            lytContent.showKeyboard();// 显示键盘
        }

        if (view.getId() == R.id.private_chat_send_btn) {
            sendType = "text";
            sendContent = chatContent.getText().toString();
            presenter.sendPrivateMsg(
                    sendContent,
                    sendType, hisId, this );
        }
    }

    @Override
    protected void setOnItemClickListener() {
        privateChatAdapter.setOnItemChildClickListener((adapter, view, position) -> {
            if (view.getId() == R.id.item_private_chat_his_img || view.getId() == R.id.item_private_chat_mine_img) {
                List<String> urls = new ArrayList<>();
                urls.add(privateChatAdapter.getData().get(position).content);
                ImageUtil.showImages(this, urls, 0);
            }
        });

        privateChatAdapter.setOnItemLongClickListener((adapter, view, position) -> {
            presenter.showDeletePrivateMsgDialog(PrivateChatActivity.this,
                    privateChatAdapter.getData().get(position).mid, hisId, position);
            return true;
        });

    }

    @Override
    public void onGetPrivateListSuccess(PrivateChatBean privateChatBean) {
        privateChatAdapter.setHisInfo(privateChatBean.body.pmList.get(0).name,
                privateChatBean.body.pmList.get(0).avatar,
                privateChatBean.body.pmList.get(0).fromUid);
        recyclerView.scheduleLayoutAnimation();
        privateChatAdapter.setNewData(privateChatBean.body.pmList.get(0).msgList);
    }

    @Override
    public void onGetPrivateListError(String msg) {
        showToast(msg, ToastType.TYPE_ERROR);
    }

    @Override
    public void onSendPrivateChatMsgSuccess(SendPrivateMsgResultBean sendPrivateMsgResultBean) {
        chatContent.setText("");

        privateChatAdapter.insertMsg(this, sendContent, sendType);
        recyclerView.scrollToPosition(privateChatAdapter.getData().size() - 1);
        showToast(sendPrivateMsgResultBean.head.errInfo, ToastType.TYPE_SUCCESS);
    }

    @Override
    public void onSendPrivateChatMsgError(String msg) {
        showToast(msg, ToastType.TYPE_ERROR);
    }

    @Override
    public void onCompressImageSuccess(List<File> compressedFiles) {
        presenter.uploadImages(compressedFiles, "pm", "image", this);
    }

    @Override
    public void onCompressImageFail(String msg) {
        showToast(msg, ToastType.TYPE_ERROR);
    }

    @Override
    public void onUploadSuccess(UploadResultBean uploadResultBean) {
        sendType = "image";
        sendContent = uploadResultBean.body.attachment.get(0).urlName;
        presenter.sendPrivateMsg(sendContent, sendType, hisId, this);
    }

    @Override
    public void onUploadError(String msg) {
        showToast(msg, ToastType.TYPE_ERROR);
    }

    @Override
    public void onDeleteSinglePmSuccess(String msg, int position) {
        privateChatAdapter.deleteMsg(position);
        ToastUtil.showToast(this, msg, ToastType.TYPE_SUCCESS);
    }

    @Override
    public void onDeleteSinglePmError(String msg) {
        showToast(msg, ToastType.TYPE_ERROR);
    }

    @Override
    public void showMsg(String msg) {
        showToast(msg, ToastType.TYPE_NORMAL);
    }

    @Override
    protected boolean registerEventBus() {
        return true;
    }

    @Override
    protected void receiveEventBusMsg(BaseEvent baseEvent) {
        if (baseEvent.eventCode == BaseEvent.EventCode.INSERT_EMOTION) {
            presenter.insertEmotion(this, chatContent, (String) baseEvent.eventData);
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == PictureConfig.CHOOSE_REQUEST) {
            List<LocalMedia> selectList = PictureSelector.obtainMultipleResult(data);
            List<String> files = new ArrayList<>();
            for (int i = 0; i < selectList.size(); i ++) {
                files.add(selectList.get(i).getRealPath());
            }
            presenter.checkBeforeSendImage(this, files);
        }
    }
}

File: app/src/main/java/com/scatl/uestcbbs/module/message/view/PrivateChatActivity.kt
package com.scatl.uestcbbs.module.message.view

import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.GradientDrawable
import android.os.CountDownTimer
import android.view.HapticFeedbackConstants
import android.view.View
import android.view.animation.AnimationUtils
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.luck.picture.lib.PictureSelector
import com.luck.picture.lib.config.PictureConfig
import com.luck.picture.lib.config.PictureMimeType
import com.scatl.uestcbbs.R
import com.scatl.uestcbbs.annotation.ToastType
import com.scatl.uestcbbs.base.BaseVBActivity
import com.scatl.uestcbbs.databinding.ActivityPrivateChatBinding
import com.scatl.uestcbbs.entity.PrivateChatBean
import com.scatl.uestcbbs.entity.PrivateChatDraftBean
import com.scatl.uestcbbs.entity.SendPrivateMsgResultBean
import com.scatl.uestcbbs.entity.UploadResultBean
import com.scatl.uestcbbs.helper.glidehelper.GlideEngineForPictureSelector
import com.scatl.uestcbbs.module.message.adapter.PrivateChatAdapter
import com.scatl.uestcbbs.module.message.presenter.PrivateChatPresenter
import com.scatl.uestcbbs.module.user.view.UserDetailActivity
import com.scatl.uestcbbs.util.Constant
import com.scatl.uestcbbs.util.ImageUtil
import com.scatl.uestcbbs.util.SharePrefUtil
import com.scatl.uestcbbs.util.showToast
import com.scatl.util.ColorUtil
import com.scatl.util.ScreenUtil
import com.scatl.widget.emotion.IEmotionEventListener
import org.litepal.LitePal
import java.io.File

/**
 * Created by sca_tl at 2023/3/29 15:35
 */
class PrivateChatActivity: BaseVBActivity<PrivateChatPresenter, PrivateChatView, ActivityPrivateChatBinding>(), PrivateChatView,
    IEmotionEventListener {

    private lateinit var privateChatAdapter: PrivateChatAdapter
    private var hisName = ""
    private var hisUid = Int.MAX_VALUE
    private var tmpContent = ""

    private val countDownTimer = object : CountDownTimer(15000, 1000) {

        override fun onTick(l: Long) {
            mBinding.edittext.isEnabled = false
            mBinding.edittext.hint = "${l / 1000}秒后可再次发消息"
        }

        override fun onFinish() {
            enableInput()
            if (tmpContent.isNotEmpty()) {
                mBinding.edittext.setText(tmpContent)
            }
        }
    }

    override fun getViewBinding() = ActivityPrivateChatBinding.inflate(layoutInflater)

    override fun initPresenter() = PrivateChatPresenter()

    override fun getContext() = this

    override fun getIntent(intent: Intent?) {
        hisName = intent?.getStringExtra(Constant.IntentKey.USER_NAME) ?: ""
        hisUid = intent?.getIntExtra(Constant.IntentKey.USER_ID, Int.MAX_VALUE) ?: Int.MAX_VALUE
    }

    override fun initView(theftProof: Boolean) {
        super.initView(true)
        mBinding.toolbar.title = hisName
        mBinding.edittext.background = GradientDrawable().apply {
            cornerRadius = ScreenUtil.dip2pxF(getContext(), 20f)
            setColor(ColorUtil.getAttrColor(getContext(), R.attr.colorSurfaceVariant))
        }

        privateChatAdapter = PrivateChatAdapter(R.layout.item_private_chat)
        privateChatAdapter.setHasStableIds(true)
        mBinding.recyclerView.apply {
            adapter = privateChatAdapter
            layoutAnimation = AnimationUtils.loadLayoutAnimation(context, R.anim.layout_animation_scale_in)
            layoutManager = LinearLayoutManager(context).apply {
                stackFromEnd = true
            }
            addItemDecoration(object : RecyclerView.ItemDecoration() {
                override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView, state: RecyclerView.State) {
                    outRect.bottom = ScreenUtil.dip2px(context, 20f)
                    outRect.left = ScreenUtil.dip2px(context, 10f)
                    outRect.right = ScreenUtil.dip2px(context, 10f)
                }
            })
        }

        bindClickEvent(mBinding.addPhotoBtn, mBinding.addEmotionBtn, mBinding.sendMsgBtn, mBinding.edittext)
        mBinding.emotionLayout.eventListener = this

        initDraft()

        mPresenter?.getPrivateMsg(hisUid)
        mPresenter?.getUserSpace(hisUid)
    }

    private fun initDraft() {
        val draft = LitePal
            .where("hostUid = ? and chatUid = ?", SharePrefUtil.getUid(getContext()).toString(), hisUid.toString())
            .find(PrivateChatDraftBean::class.java)
        if (draft.isNotEmpty()) {
            mBinding.edittext.setText(draft[0].content)
        }
    }

    override fun onClick(v: View) {
        super.onClick(v)
        when(v) {
            mBinding.addPhotoBtn -> {
                v.performHapticFeedback(HapticFeedbackConstants.CONTEXT_CLICK)
                PictureSelector
                    .create(this)
                    .openGallery(PictureMimeType.ofImage())
                    .isCamera(true)
                    .isGif(true)
                    .showCropFrame(false)
                    .hideBottomControls(false)
                    .theme(com.luck.picture.lib.R.style.picture_WeChat_style)
                    .maxSelectNum(1)
                    .isEnableCrop(false)
                    .imageEngine(GlideEngineForPictureSelector.createGlideEngine())
                    .forResult(PictureConfig.CHOOSE_REQUEST)
            }
            mBinding.addEmotionBtn -> {
                v.performHapticFeedback(HapticFeedbackConstants.CONTEXT_CLICK)
                if (mBinding.emotionLayout.visibility == View.GONE) {
                    mBinding.smoothInputLayout.showInputPane(true)
                    mBinding.emotionLayout.visibility = View.VISIBLE
                } else {
                    mBinding.smoothInputLayout.showKeyboard()
                }
            }
            mBinding.edittext -> {
                mBinding.smoothInputLayout.showKeyboard()
            }
            mBinding.sendMsgBtn -> {
                disableInput()
                mPresenter?.sendPrivateMsg(mBinding.edittext.text.toString(), "text", hisUid)
            }
        }
    }

    override fun setOnItemClickListener() {
        privateChatAdapter.setOnItemChildClickListener { adapter, view, position ->
            when(view.id) {
                R.id.his_img_content, R.id.mine_img_content -> {
                    val urls = ArrayList<String>()
                    urls.add(privateChatAdapter.data[position].content)
                    ImageUtil.showImages(this, urls, 0)
                }
                R.id.his_icon, R.id.mine_icon -> {
                    val intent = Intent(getContext(), UserDetailActivity::class.java).apply {
                        putExtra(Constant.IntentKey.USER_ID, privateChatAdapter.data[position].sender)
                    }
                    startActivity(intent)
                }
            }
        }

        privateChatAdapter.setOnItemLongClickListener { adapter, view, position ->
            mPresenter?.showDeletePrivateMsgDialog(privateChatAdapter.data[position].mid, hisUid, position)
            true
        }
    }

    override fun onGetPrivateListSuccess(privateChatBean: PrivateChatBean) {
        mBinding.recyclerView.scheduleLayoutAnimation()
        privateChatAdapter.setNewData(privateChatBean.body.pmList[0].msgList)
    }

    override fun onGetPrivateListError(msg: String?) {
        showToast(msg, ToastType.TYPE_ERROR)
    }

    override fun onSendPrivateChatMsgSuccess(sendPrivateMsgResultBean: SendPrivateMsgResultBean, content: String?, type: String) {
        disableInput()
        tmpContent = if ("image" == type) {
            mBinding.edittext.text.toString()
        } else {
            ""
        }
        mBinding.edittext.setText("")
        countDownTimer.cancel()
        countDownTimer.start()

        privateChatAdapter.insertMsg(this, content, type)
        mBinding.recyclerView.scrollToPosition(privateChatAdapter.data.size - 1)
        showToast(sendPrivateMsgResultBean.head.errInfo, ToastType.TYPE_SUCCESS)
    }

    override fun onSendPrivateChatMsgError(msg: String?) {
        enableInput()
        showToast(msg, ToastType.TYPE_ERROR)
    }

    override fun onCompressImageSuccess(compressedFiles: List<File>) {
        disableInput()
        mPresenter?.uploadImages(compressedFiles, "pm", "image")
    }

    override fun onCompressImageFail(msg: String?) {
        enableInput()
        showToast(msg, ToastType.TYPE_ERROR)
    }

    override fun onUploadSuccess(uploadResultBean: UploadResultBean) {
        disableInput()
        mPresenter?.sendPrivateMsg(uploadResultBean.body.attachment.getOrNull(0)?.urlName, "image", hisUid)
    }

    override fun onUploadError(msg: String?) {
        enableInput()
        showToast(msg, ToastType.TYPE_ERROR)
    }

    override fun onDeleteSinglePmSuccess(msg: String?, position: Int) {
        privateChatAdapter.deleteMsg(position)
        showToast(msg, ToastType.TYPE_SUCCESS)
    }

    override fun onDeleteSinglePmError(msg: String?) {
        showToast(msg, ToastType.TYPE_ERROR)
    }

    override fun onGetUserSpaceSuccess(isOnline: Boolean) {
        mBinding.onlineStatus.backgroundTintList =
            if (isOnline) {
                ColorStateList.valueOf(Color.parseColor("#FF049E3B"))
            } else {
                ColorStateList.valueOf(ColorUtil.getAttrColor(this, R.attr.colorOutline))
            }
    }

    override fun onGetUserSpaceError(msg: String?) {

    }

    private fun disableInput() {
        mBinding.edittext.isEnabled = false
        mBinding.edittext.hint = "消息发送中..."
        mBinding.addEmotionBtn.isEnabled = false
        mBinding.addPhotoBtn.isEnabled = false
        mBinding.sendMsgBtn.isEnabled = false
    }

    private fun enableInput() {
        mBinding.edittext.isEnabled = true
        mBinding.edittext.hint = "请在此输入消息内容~"
        mBinding.addEmotionBtn.isEnabled = true
        mBinding.addPhotoBtn.isEnabled = true
        mBinding.sendMsgBtn.isEnabled = true
    }

    override fun onEmotionClick(path: String?) {
        if (!mBinding.edittext.isEnabled) {
            showToast("请稍候...", ToastType.TYPE_NORMAL)
            return
        }
        path?.let {
            mBinding.edittext.insertEmotion(path)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == PictureConfig.CHOOSE_REQUEST) {
            val selectList = PictureSelector.obtainMultipleResult(data)
            val files: MutableList<String> = java.util.ArrayList()
            for (i in selectList.indices) {
                files.add(selectList[i].realPath)
            }
            mPresenter?.checkBeforeSendImage(files)
        }
    }

    override fun onStop() {
        saveDraft()
        super.onStop()
    }

    private fun saveDraft() {
        val draft = PrivateChatDraftBean().apply {
            hostUid = SharePrefUtil.getUid(getContext())
            chatUid = hisUid
            content = mBinding.edittext.text.toString()
        }
        draft.saveOrUpdate("hostUid = ? and chatUid = ?", SharePrefUtil.getUid(getContext()).toString(), hisUid.toString())
    }

    override fun onDestroy() {
        super.onDestroy()
        countDownTimer.cancel()
    }
}